### PR TITLE
RFC: AMP versioning policy

### LINKED ISSUE #5160: Discussion - do we need a breaking changes policy?
Since changes to the AMP runtime can have a profound impact of existing content there should be policy around breaking changes.  Specifically:
1.  How is a breaking change defined? 
2. What is the process for notification?

**With that in mind here is a straw man policy for discussion**

A breaking change is any change that will cause a existing AMP pages to render in way that is one of:
1. Noticeably, visually, different,  
2. Where non visual functionality will not longer produce the same external action.

Example of case 1 would be Issue #5152.  An example of case 2 would be changing validation to make `amp-ad.js` mandatory which would break all existing ads.  Another example of case 2: changes to `amp-analytics` that causes existing scripts to no longer report the same data.

Breaking changes should be subject to a notification policy (beyond the normal github PR) email.   This should include:
1. An email "announce" list that all content developers can subscribe to.
2. At least 14 days notice of a change.
3. Validator warnings for a similar period.
4. A clear, scheduled, deployment date to both experiment and prod.
5. Post deployment confirmation notification.

Exception - any change that fixes a bug to bring behavior into conformance with existing documentation or restore existing behavior in response to browser bugs is, at the discretion of the Tech Lead,  not subject to the 14 days advance notice requirement.

### PR DESCRIPTION
**Pull request by cramforce**
Fixes #5160

### CODE REVIEW

File: spec/amp-versioning-policy.md [hunk 1]
```diff
@@ -0,0 +1,40 @@
+# AMP versioning policy
+
+## Basics
+
+- We use [semver](http://semver.org/) terminology, but see below for changes from simple semver.
+- The AMP JS library is distributed via its major version only. The major version is expected to change rarely if ever (on the order of less than once per year).
+- Each AMP extension is versioned independently. Publishers of AMP documents can choose the major and minor version of an extension. They cannot choose the patch version.
```
Comment (erwinmombay) [inline spec/amp-versioning-policy.md]: Do we guarantee compatibility with the core binary?

**Comment by erwinmombay** [comment]
cc @jpettitt @dknecht @jrf0110

File: spec/amp-versioning-policy.md [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+- AMP is **released** about once per weak and each release has a unique version number. With each release, AMP pages are rendered with the latest code for the respective major/minor versions of the AMP JS library and extensions.
+
+## Goals
+
+- Reduce overall JS URL entropy. AMP aims to provide as few versions of the JS library and extensions as possible to improve cache hit rates for these requests.
+- Avoid breaking existing pages.
+- Keep all users of AMP on secure versions of the library.
```
Comment (erwinmombay) [inline spec/amp-versioning-policy.md]: How do we achieve this if the publisher has been pinned to an old version and a major change was required for a security fix

File: spec/amp-versioning-policy.md [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+Breaking changes are:
+
+- Changes to the semantics of supported attributes and other exposed APIs.
+- Changes to CSS included in AMP that are not backward compatible.
+- Changes to the DOM structure (elements, attributes and attribute values) generated by AMP extensions that are not backward compatible.
+
+The following special cases are not considered breaking changes:
+
+- Changes to elements and their children if the element name starts with `i-`.
+- Changes to attribute names starting with `-`.
```
Comment (erwinmombay) [inline spec/amp-versioning-policy.md]: maybe add classes too

File: spec/amp-versioning-policy.md (hunk 2 above)
Comment (cramforce) [inline spec/amp-versioning-policy.md]: Back port to each old version.

File: spec/amp-versioning-policy.md (hunk 1 above)
Comment (erwinmombay) [inline spec/amp-versioning-policy.md]: cool. can this supported by validator?

cc @powdercloud @Gregable

File: spec/amp-versioning-policy.md [hunk 4]
```diff
@@ -0,0 +1,40 @@
+# AMP versioning policy
+
+## Basics
+
+- We use [semver](http://semver.org/) terminology, but see below for changes from simple semver.
+- The AMP JS library is distributed via its major version only. The major version is expected to change rarely if ever (on the order of less than once per year).
+- Each AMP extension is versioned independently. Publishers of AMP documents can choose the major and minor version of an extension. They cannot choose the patch version.
+- AMP is **released** about once per weak and each release has a unique version number. With each release, AMP pages are rendered with the latest code for the respective major/minor versions of the AMP JS library and extensions.
```
Comment (erwinmombay) [inline spec/amp-versioning-policy.md]: yeah release vs version SGTM

File: spec/amp-versioning-policy.md (hunk 2 above)
Comment (erwinmombay) [inline spec/amp-versioning-policy.md]: 👍

File: spec/amp-versioning-policy.md (hunk 3 above)
Comment (jpettitt) [inline spec/amp-versioning-policy.md]: Not sure how to phrase it but there is potential for some conflict between the '-' and 'i-' exceptions and the non backward compatible CSS provision, particularly if you add -amp classes to the list as @erwinmombay suggests.  Which one rules?  Perhaps a definition if the scope of i- and - attributes and classes is needed someplace else?

File: spec/amp-versioning-policy.md [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+- Changes to elements and their children if the element name starts with `i-`.
+- Changes to attribute names starting with `-`.
+- Changes required to maintain the security of AMP pages.
+- Changes that break fewer than 0.1% of crawler accessible AMP pages.
+
+For breaking changes the major version of an extension (or more rarely) the main AMP JS library should be increased.
+
+## Removing old versions
+
+Old versions of extensions may be removed if they are used by fewer than 0.1% of crawler accessible AMP pages.
```
Comment (jpettitt) [inline spec/amp-versioning-policy.md]: Can you define a notification policy/mechanism breaking changes and for the 0.1% cases ?  Security case?

File: spec/amp-versioning-policy.md (hunk 3 above)
Comment (cramforce) [inline spec/amp-versioning-policy.md]: Will do. The backward compat provision wins, except when CSS selectors match these elements that should not be hit by user CSS.

File: spec/amp-versioning-policy.md (hunk 5 above)
Comment (cramforce) [inline spec/amp-versioning-policy.md]: Very good point. Will do.

File: spec/amp-versioning-policy.md (hunk 1 above)
Comment (cramforce) [inline spec/amp-versioning-policy.md]: I don't think the validator has a role in this. This just says how the versions themselves are composed.

File: spec/amp-versioning-policy.md (hunk 4 above)
Comment (cramforce) [inline spec/amp-versioning-policy.md]: Added sentence at the top

File: spec/amp-versioning-policy.md (hunk 3 above)
Comment (cramforce) [inline spec/amp-versioning-policy.md]: I clarified this.

File: spec/amp-versioning-policy.md (hunk 5 above)
Comment (cramforce) [inline spec/amp-versioning-policy.md]: Done. Also first draft of deprecation policy.

File: spec/amp-versioning-policy.md (hunk 1 above)
Comment (erwinmombay) [inline spec/amp-versioning-policy.md]: I guess im confused where the declaration will be, is it just by convention or some explicit assertion somewhere

**Comment by cramforce** [comment]
Ping for LG or additional feedback :)

File: spec/amp-versioning-policy.md (hunk 4 above)
Comment (rosariomgomez) [inline spec/amp-versioning-policy.md]: small typo: "once per weak" should be "once per week"

**Comment by jrf0110** [comment]
Meant to say after the original ping, everything LGTM 👍  One thing I was a little curious about was this line:

> If a change is backward compatible, no version should be changed.

How does that work? My confusion here might just be from my inexperience with the codebase. Does the Major/minor combo get re-tagged? Or is there a patch bump?

**Comment by cramforce** [comment]
@jrf0110 AMP releases once a week. These releases can make changes to AMP in accordance with this policy that may not affect versioning. In particular patch versions (which largely equivalent to releases are never exposed to users.)

**Comment by jpettitt** [comment]
I think the issues we had with fixed layer / amp-sidebar this week highlight a weakness in the underlying design.  

Since AMP is sharing the CSS space with publisher CSS the opportunities for unintended side effect breaking things are essentially limitless.  While I'm all in favor of TDD/CI/CD the TDD part is impossible to do with any degree of certainty in this model.  I'm not sure what the solution is.  

With 29,000 github emails from the _amproject_ or _amphtml_ in the last year it's ~~hard~~ impossible to track every change for possible conflicts with our code. And that's with a company dedicated 100% full time to AMP, a regular publisher who "does an AMP version" then moves on doesn't stand a chance.

Perhaps it's time to move to a [_semver_](http://semver.org/) model?

**Comment by jpettitt** [comment]
If you can make the visual diffing work that would do it.  I'm skeptical.

**Comment by cramforce** [comment]
We already use that quite a bit for optimizations on the cache side. Will need some work, but should help a lot.

The other thing we'll do (as 2017 moves on) is focus on making it much more natural to introduce new versions of extensions. Right now this involves too much work on the engineer side, so people try to avoid it at all cost which is not healthy.

**Pull request merged**